iceoryx_posh/source/popo/waitset/wait_set.cpp

```diff
@@ -0,0 +1,125 @@
+// Copyright (c) 2020 by Robert Bosch GmbH. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License
+
+#include "iceoryx_posh/internal/popo/waitset/wait_set.hpp"
+
+namespace iox
+{
+namespace popo
+{
+WaitSet::WaitSet(cxx::not_null<ConditionVariableData* const> condVarDataPtr) noexcept
+    : m_conditionVariableDataPtr(condVarDataPtr)
+    , m_conditionVariableWaiter(m_conditionVariableDataPtr)
+    , m_guardCondition(m_conditionVariableDataPtr)
+{
+    m_conditionVector.push_back(&m_guardCondition);
+}
+
+bool WaitSet::attachCondition(Condition& condition) noexcept
+{
+    if (!condition.isConditionVariableAttached())
+    {
+        if (condition.attachConditionVariable(m_conditionVariableDataPtr))
+        {
+            return m_conditionVector.push_back(&condition);
+        }
+    }
+
+    return false;
+}
+
+bool WaitSet::detachCondition(const Condition& condition) noexcept
+{
+    for (auto currentCondition : m_conditionVector)
+    {
+        if (currentCondition == &condition)
+        {
+            m_conditionVector.erase(&currentCondition);
+            return true;
+        }
+    }
+    return false;
+}
+
+void WaitSet::clear() noexcept
+{
+    m_conditionVector.clear();
+}
+
+cxx::vector<Condition*, MAX_NUMBER_OF_CONDITIONS>
+WaitSet::waitAndReturnFulfilledConditions(cxx::optional<units::Duration> timeout) noexcept
+{
+    cxx::vector<Condition*, MAX_NUMBER_OF_CONDITIONS> conditionsWithFulfilledPredicate;
+
+    /// @note Inbetween here and last wait someone could have set the trigger to true, hence reset it
+    m_conditionVariableWaiter.reset();
+
+    // Is one of the conditons true?
+    for (auto currentCondition : m_conditionVector)
```

Comment:
This should be `auto & currentCondition : m_conditionVector` otherwise we would work on copies. Actually with the current `cxx::vector` implementation this is correct code but this is due to a simplified iterator in this vector which should be fixed sooner or later.

Reply:
Good to know, done